controller/gateway/controller_cleanup.go

```diff
@@ -129,3 +133,28 @@ func (r *Reconciler) cleanup(
 	log.Debug(logger, "owned resources cleanup completed", gateway)
 	return true, ctrl.Result{}, nil
 }
+
+func handleGatewayFinalizerPatchOrUpdateError(err error, gateway *gatewayv1.Gateway, logger logr.Logger) (ctrl.Result, error) {
+	// If the Gateway is not found, then requeue without an error.
+	if k8serrors.IsNotFound(err) {
+		return ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: requeueWithoutBackoff,
+		}, nil
+	}
+	// Since controllers use cached clients, it's possible that the Gateway is out of sync with what
+	// is in the API server and this causes:
+	// Forbidden: no new finalizers can be added if the object is being deleted, found new finalizers []string{...}
+	// Code below handles that gracefully to not show users the errors that are not actionable.
+	if cause, ok := k8serrors.StatusCause(err, metav1.CauseTypeForbidden); k8serrors.IsInvalid(err) && ok {
+		log.Debug(logger, "failed to delete a finalizer on Gateway, requeueing request", gateway, "cause", cause)
+		return ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: requeueWithoutBackoff,
+		}, nil
+	}
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
```

Comment:
```suggestion
	return ctrl.Result{}, err
```

btw. can we shortcircuit at the top with
`if err == nil`?

Reply:
This way it's consistent that the least indented path is a happy path + the explicit `nil` makes it easy to figure out.

I'm OK with accepting the suggestion if you fill like this would improve the readability on your end. I might be biased.

Reply:
Both have benefits and drawbacks, this can be left as it is. 

Btw. putting `if err == nil` to return at the top?

Reply:
4d9f6e6b025529ea01f69dc3bb62c6481373236e

👍